python/lsst/obs/metadata/metadata.py

```diff
@@ -0,0 +1,179 @@
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Classes and support code for metadata translation"""
+
+from abc import abstractmethod
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class MetadataMeta(type):
+    """Register all subclasses with the base class"""
+
+    @staticmethod
+    def _makeConstMapping(standardKey, constant):
+        def translator(header):
+            return constant
+        translator.__doc__ = f"""Returns constant value for '{standardKey}' property
+
+        Parameters
+        ----------
+        header : `dict`-like
+            Header values representing a FITS header.
+
+        Returns
+        -------
+        translation : `str` or `numbers.Number`
+            Always returns {constant}.
+        """
+        return translator
+
+    @staticmethod
+    def _makeUnitMapping(standardKey, fitsKey):
+        def translator(header):
+            return header[fitsKey]
+        translator.__doc__ = f"""Map '{fitsKey}' FITS keyword to '{standardKey}' property
+
+        Parameters
+        ----------
+        header : `dict`-like
+            Header values representing a FITS header.
+
+        Returns
+        -------
+        translation : `str` or `numbers.Number`
+            Translated header value derived directly from the {fitsKey}
+            header value.
+        """
+        return translator
+
+    def __init__(self, name, bases, dct):
+        super().__init__(name, bases, dct)
+
+        if hasattr(self, "name") and self.name is not None:
+            MetadataTranslator.translators[self.name] = self
+
+        for standardKey, fitsKey in self._unitMap.items():
+            translator = self._makeUnitMapping(standardKey, fitsKey)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+        for standardKey, constant in self._constMap.items():
+            translator = self._makeConstMapping(standardKey, constant)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+
+class MetadataTranslator(metaclass=MetadataMeta):
+    """Per-instrument metadata translation support"""
+
+    _unitMap = {}
+    """Dict of one-to-one mappings for header translation from standard
+    property to corresponding FITS keyword."""
```

Comment:
Using a leading-underscore class variable as a customization point bothers me slightly.  It's partly the old "is `protected` more like `private` or more like `public` question, but only in part; I also want something how that somehow looks like a customization hook, the way an abstract method does.  But I don't actually have an obviously better suggestion, and I don't want to hold up actual functionality by bikeshedding the metaprogramming, so let's not worry about it for now.  But maybe keep it in the back of your mind.

Reply:
Yes. The dict here is only used by the subclasses and the metaclass and nowhere else. I should almost set it to None after I've created the methods dynamically. I have to put it somewhere and it's clearly not public.  I do like how the metaclasses work such that sphinx/pydoc show the documentation with all the methods in place.